File: android/siopkit/src/main/kotlin/jp/co/pendako/siop/RsaPublicJwk.kt
package jp.co.pendako.siop

import java.math.BigInteger
import java.security.KeyFactory
import java.security.MessageDigest
import java.security.PublicKey
import java.security.interfaces.RSAPublicKey
import java.security.spec.RSAPublicKeySpec

/** RSA public key as a JWK (RFC 7517), as carried in the `sub_jwk` claim. */
data class RsaPublicJwk(val n: String, val e: String) {
    val kty: String get() = "RSA"

    /**
     * JWK thumbprint (RFC 7638): SHA-256 over the canonical JSON of the
     * required members in lexicographic order. This is the `sub` of a
     * self-issued ID Token (OpenID Connect Core 1.0 Section 7.4).
     */
    fun thumbprint(): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(canonicalJson.toByteArray())
        return Base64Url.encode(digest)
    }

    /** The JSON the thumbprint is taken over: the required members, in order, with no whitespace. */
    val canonicalJson: String get() = """{"e":"$e","kty":"$kty","n":"$n"}"""

    fun toJsonObject(): Map<String, String> = mapOf("kty" to kty, "n" to n, "e" to e)

    fun toPublicKey(): PublicKey {
        val modulus = BigInteger(1, Base64Url.decode(n))
        val exponent = BigInteger(1, Base64Url.decode(e))
        return KeyFactory.getInstance("RSA").generatePublic(RSAPublicKeySpec(modulus, exponent))
    }

    companion object {
        fun from(key: RSAPublicKey): RsaPublicJwk = RsaPublicJwk(
            n = Base64Url.encode(key.modulus.toUnsignedBytes()),
            e = Base64Url.encode(key.publicExponent.toUnsignedBytes()),
        )

        /** Big-endian magnitude without the sign byte BigInteger prepends. */
        private fun BigInteger.toUnsignedBytes(): ByteArray {
            val bytes = toByteArray()
            return if (bytes.size > 1 && bytes[0] == 0.toByte()) bytes.copyOfRange(1, bytes.size) else bytes
        }
    }
}

File: ios/SIOPKit/Sources/SIOPKit/RSAPublicJWK.swift
import CryptoKit
import Foundation
import Security

/// RSA public key as a JWK (RFC 7517), as embedded in the `sub_jwk` claim.
public struct RSAPublicJWK: Equatable {
    public let n: String
    public let e: String
    public var kty: String { "RSA" }

    public init(n: String, e: String) {
        self.n = n
        self.e = e
    }

    public var jsonObject: [String: String] {
        ["kty": kty, "n": n, "e": e]
    }

    /// RFC 7638 JWK thumbprint (SHA-256, base64url) — used as the `sub` claim
    /// value of a self-issued ID Token (OpenID Connect Core 1.0 Section 7.4).
    public func thumbprint() -> String {
        let canonical = "{\"e\":\"\(e)\",\"kty\":\"\(kty)\",\"n\":\"\(n)\"}"
        let digest = SHA256.hash(data: Data(canonical.utf8))
        return Base64URL.encode(Data(digest))
    }

    /// Reconstructs a Security-framework public key, e.g. for signature verification.
    public func secKey() throws -> SecKey {
        guard let nData = Base64URL.decode(n), let eData = Base64URL.decode(e) else {
            throw SIOPError.invalidKey
        }
        let der = DER.encodeRSAPublicKey(n: nData, e: eData)
        let attributes: [String: Any] = [
            kSecAttrKeyType as String: kSecAttrKeyTypeRSA,
            kSecAttrKeyClass as String: kSecAttrKeyClassPublic,
        ]
        var error: Unmanaged<CFError>?
        guard let key = SecKeyCreateWithData(der as CFData, attributes as CFDictionary, &error) else {
            throw SIOPError.invalidKey
        }
        return key
    }
}
